internal/web3_provider/permits.go

```diff
@@ -0,0 +1,188 @@
+package web3_provider
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/signer/core/apitypes"
+
+	"github.com/1inch/1inch-sdk-go/internal/helpers/consts/abis"
+)
+
+type ContractPermitData struct {
+	FromToken     string
+	Spender       string
+	Name          string
+	Version       string
+	PublicAddress string
+	ChainId       int
+	Key           string
+	Nonce         int64
+	Deadline      int64
+	Amount        string
+}
+
+// TokenPermit Will return an erc2612 string struct if possible
+func (w Wallet) TokenPermit(cd ContractPermitData) (string, error) {
+	domainData := apitypes.TypedDataDomain{
+		Name:              cd.Name,
+		Version:           cd.Version,
+		ChainId:           math.NewHexOrDecimal256(int64(cd.ChainId)),
+		VerifyingContract: cd.FromToken,
+	}
+
+	// Order Message
+	orderMessage := apitypes.TypedDataMessage{
+		"owner":    cd.PublicAddress,
+		"spender":  cd.Spender,
+		"value":    cd.Amount,
+		"nonce":    big.NewInt(cd.Nonce),
+		"deadline": big.NewInt(cd.Deadline),
+	}
+
+	// Typed Data
+	typedData := apitypes.TypedData{
+		Types: map[string][]apitypes.Type{
+			"EIP712Domain": {
+				{Name: "name", Type: "string"},
+				{Name: "version", Type: "string"},
+				{Name: "chainId", Type: "uint256"},
+				{Name: "verifyingContract", Type: "address"},
+			},
+			"Permit": {
+				{Name: "owner", Type: "address"},
+				{Name: "spender", Type: "address"},
+				{Name: "value", Type: "uint256"},
+				{Name: "nonce", Type: "uint256"},
+				{Name: "deadline", Type: "uint256"},
+			},
+		},
+		PrimaryType: "Permit",
+		Domain:      domainData,
+		Message:     orderMessage,
+	}
+
+	typedDataHash, err := typedData.HashStruct(typedData.PrimaryType, typedData.Message)
```

Comment:
we need to use multicall to optimize amount of call to the NODE RPC 

@Tanz0rz will be grateful if you can help